### PR TITLE
dma: stm32 driver cleanup

### PR DESCRIPTION
**Pull request by erwango**
This change is an extract of #23157 focused on changes that do not impact SPI

EDIT by avisconti::
Tested on shield iks02a1 with microphone sample
(https://github.com/avisconti/zephyr/tree/test_new_dmastm32)

### CODE REVIEW

File: include/dt-bindings/dma/stm32_dma.h [hunk 1]
```diff
@@ -1,16 +1,38 @@
...
+/* 0 -> MEM_TO_MEM, 1 -> MEM_TO_PERIPH, 2 -> PERIPH_TO_MEM */
+#define STM32_DMA_CONFIG_DIRECTION(config)		((config >> 6) & 0x3)
+/* periph increment defined on bit 9 as true/false */
+#define STM32_DMA_CONFIG_PERIPH_ADDR_INC(config)	((config >> 9) & 0x1)
+/* mem increment defined on bit 10 as true/false */
+#define STM32_DMA_CONFIG_MEM_ADDR_INC(config)		((config >> 10) & 0x1)
+/* perih data size defined on bits 11-12 */
+/* 0 -> 1 byte, 1 -> 2 bytes, 2 -> 4 bytes */
+#define STM32_DMA_CONFIG_PERIPH_DATA_SIZE(config)	\
+	((((config >> 11) & 0x3) == 0) ? 1 : (((config >> 11) & 0x3) * 2))
```
Comment (cybertale) [inline include/dt-bindings/dma/stm32_dma.h:21]: I think
```
#define STM32_DMA_CONFIG_PERIPH_DATA_SIZE(config) (1 << ((config >> 11) & 0x3))
```
is better, because this calculation is just `2^((config >> 11) & 0x3)`. What do you think?

File: include/dt-bindings/dma/stm32_dma.h [hunk 2]
```diff
@@ -1,16 +1,38 @@
...
+/* memory data size defined on bits 13, 14 */
+/* 0 -> 1 byte, 1 -> 2 bytes, 2 -> 4 bytes */
+#define STM32_DMA_CONFIG_MEM_DATA_SIZE(config)		\
+	((((config >> 13) & 0x3) == 0) ? 1 : (((config >> 13) & 0x3) * 2))
```
Comment (cybertale) [inline include/dt-bindings/dma/stm32_dma.h:25]: Same as above.

File: drivers/dma/dma_stm32.c [hunk 3]
```diff
@@ -198,20 +201,26 @@ static int dma_stm32_get_periph_increment(enum dma_addr_adj increment,
 static int dma_stm32_configure(struct device *dev, u32_t id,
 			       struct dma_config *config)
 {
+
```
Comment (cybertale) [inline drivers/dma/dma_stm32.c:216]: Why this empty line?

File: drivers/dma/dma_stm32.c [hunk 4]
```diff
@@ -198,20 +201,26 @@ static int dma_stm32_get_periph_increment(enum dma_addr_adj increment,
...
 	struct dma_stm32_data *data = dev->driver_data;
-	struct dma_stm32_stream *stream = &data->streams[id];
 	const struct dma_stm32_config *dev_config =
 					dev->config->config_info;
 	DMA_TypeDef *dma = (DMA_TypeDef *)dev_config->base;
+	struct dma_stm32_stream *stream = &data->streams[id - STREAM_OFFSET];
```
Comment (cybertale) [inline drivers/dma/dma_stm32.c:221]: Out of curiosity, why are we moving this line? General idea of initializations should be following the reverse christmas tree format, which means the long lines should go up rather than down.

File: drivers/dma/dma_stm32.c [hunk 5]
```diff
@@ -198,20 +201,26 @@ static int dma_stm32_get_periph_increment(enum dma_addr_adj increment,
...
 	LL_DMA_InitTypeDef DMA_InitStruct;
 	u32_t msize;
 	int ret;
 
+	/* give channel from index 0 */
+	id = id - STREAM_OFFSET;
+
 	if (id >= data->max_streams) {
+		LOG_ERR("cannot configure the dma stream %d.", id);
```
Comment (cybertale) [inline drivers/dma/dma_stm32.c:230]: Maybe here we use "channel index setup is not valid with value %d" is better?

File: drivers/dma/dma_stm32.c [hunk 6]
```diff
@@ -304,7 +313,6 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 		periph_addr_adj = config->head_block->dest_addr_adj;
 		break;
 	/* Direction has been asserted in dma_stm32_get_direction. */
-	}
```
Comment (cybertale) [inline drivers/dma/dma_stm32.c:309]: Is this a typo?

File: drivers/dma/dma_stm32.c (hunk 3 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32.c:216]: no reason, removed

File: drivers/dma/dma_stm32.c (hunk 4 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32.c:221]: Ok, (the reason is that line has been moved several time,)
Restoring initial position

File: drivers/dma/dma_stm32.c (hunk 5 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32.c:230]: OK, The fact is that this "Id" refer to a dma stream. So I propose 
LOG_ERR("Channel %d is not valid for this dma.", id);

File: drivers/dma/dma_stm32.c (hunk 6 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32.c:309]: yes, fixed elsewhere, now fixed in this commit

**Comment by erwango** [comment]
@cybertale nice to see you back, hope you are doing fine.
Could you also have a look to #23157 and #23155 which are 2 functional improvements we would like to see progressing?. This PR is essentially an abstract of these as preparation work.

**Comment by cybertale** [comment]
@erwango , thanks for the concern! Due to the reason that we all know, the COVID-19 pandemic, we still can not go back to school, where all my hardwares are. So I am not able to do any actually tests. But I do can help review the code. I'll look into PRs you mentioned.

**Comment by erwango** [comment]
@cybertale, I think I've fixed all your comments

**Comment by FRASTM** [comment]
I have introduced the dma: stm32 driver cleanup #24333 
as an equivalent PR

File: include/dt-bindings/dma/stm32_dma.h [hunk 7]
```diff
@@ -1,16 +1,38 @@
...
-#define STM32_DMA_CONFIG_PERIPHERAL_DATA_SIZE(config)	(config & (0x3 << 11))
-#define STM32_DMA_CONFIG_MEMORY_DATA_SIZE(config)	(config & (0x3 << 13))
-#define STM32_DMA_CONFIG_PERIPHERAL_INC_FIXED(config)	(config & 0x1 << 15)
+/* direction defined on bits 6-7 */
+/* 0 -> MEM_TO_MEM, 1 -> MEM_TO_PERIPH, 2 -> PERIPH_TO_MEM */
+#define STM32_DMA_CONFIG_DIRECTION(config)		((config >> 6) & 0x3)
+/* periph increment defined on bit 9 as true/false */
+#define STM32_DMA_CONFIG_PERIPHERAL_ADDR_INC(config)	((config >> 9) & 0x1)
+/* mem increment defined on bit 10 as true/false */
+#define STM32_DMA_CONFIG_MEMEMORY_ADDR_INC(config)	((config >> 10) & 0x1)
```
Comment (FRASTM) [inline include/dt-bindings/dma/stm32_dma.h:17]: STM32_DMA_CONFIG_MEMORY_ADDR_INC

Comment (erwango) [reply include/dt-bindings/dma/stm32_dma.h]: oups, txs!

**Review by FRASTM — COMMENTED** [review]
commit msg 
"Move tables declaration **as** the are only used locally for now"

**Review by FRASTM — APPROVED** [review]

**Comment by erwango** [comment]
>commit msg
"Move tables declaration as the are only used locally for now"

Fixed txs.

File: include/dt-bindings/dma/stm32_dma.h [hunk 8]
```diff
@@ -1,16 +1,38 @@
...
+/* 0 -> 1 byte, 1 -> 2 bytes, 2 -> 4 bytes */
+#define STM32_DMA_CONFIG_PERIPHERAL_DATA_SIZE(config)	\
+						(1 << ((config >> 11) & 0x3))
+/* memory data size defined on bits 13, 14 */
+/* 0 -> 1 byte, 1 -> 2 bytes, 2 -> 4 bytes */
+#define STM32_DMA_CONFIG_MEMORY_DATA_SIZE(config)	\
+						(1 << ((config >> 13) & 0x3))
+/* priority increment offset defined on bit 15 */
+#define STM32_DMA_CONFIG_PERIPHERAL_INC_FIXED(config)	((config >> 15) & 0x1)
+/* priority defined on bits 16-17 as 0, 1, 2, 3 */
```
Comment (avisconti) [inline include/dt-bindings/dma/stm32_dma.h:28]: Seems like before there was a bug, isn't it?
Or maybe the semantic of the defines is changed?

File: drivers/dma/dma_stm32.h [hunk 9]
```diff
@@ -45,6 +45,15 @@ struct dma_stm32_config {
...
 
+#ifdef CONFIG_DMA_STM32_V1
+/* from DTS the dma stream id is in range 0..<dma-requests>-1 */
+#define STREAM_OFFSET 0
+#else
+/* from DTS the dma stream id is in range 1..<dma-requests> */
+/* so decrease to set range from 0 from now on */
+#define STREAM_OFFSET 1
+#endif /* CONFIG_DMA_STM32_V1 */
+
```
Comment (avisconti) [inline drivers/dma/dma_stm32.h:44]: It seems to me that for L4 is 1 to 7 (not 8). Correct?
For F4 is 0 to 7.

File: drivers/dma/dma_stm32_v2.c [hunk 10]
```diff
@@ -117,7 +117,7 @@ static void (*func_ll_clear_te[])(DMA_TypeDef *DMAx) = {
 #endif /* LL_DMA_IFCR_CTEIF8 */
 };
 
-static void (*func_ll_clear_gi[])(DMA_TypeDef *DMAx) = {
+void (*func_ll_clear_gi[])(DMA_TypeDef *DMAx) = {
```
Comment (avisconti) [inline drivers/dma/dma_stm32_v2.c:120]: Why did you remove the static here and declared it extern?
I understand that it is gonna used only in this file. So why?

File: drivers/dma/dma_stm32.c [hunk 11]
```diff
@@ -287,6 +287,16 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 	stream->src_size	= config->source_data_size;
 	stream->dst_size	= config->dest_data_size;
 
+	if ((&config->head_block->source_address == NULL)
```
Comment (avisconti) [inline drivers/dma/dma_stm32.c:290]: What is the meaning of this check? Test if config->head_block is NULL? So, why not checking it in this way?
Moreover, if this is what you want to check, maybe is better to test it in the beginning of the routine.

File: drivers/dma/dma_stm32.c [hunk 12]
```diff
@@ -324,6 +324,10 @@ static int dma_stm32_configure(struct device *dev, u32_t id,
 		periph_addr_adj = config->head_block->dest_addr_adj;
 		break;
 	/* Direction has been asserted in dma_stm32_get_direction. */
+	default:
+		LOG_ERR("Channel direction error. %d",
```
Comment (avisconti) [inline drivers/dma/dma_stm32.c:338]: maybe you may want to get rid of the '.' here.

File: drivers/dma/dma_stm32.h (hunk 9 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32.h:44]: Yes, this is very confusing in the stm32 ref manuals.
For F4 (DMA V1) requests are called 'channels' 
For L4 (DMA V2) channels (what DMA V1 calls 'streams') have several requests.
I tried to format in the bindings the channel (or stream for DMA_V1) in the first position and the request (or channel in DMA_V1) in the second position, based on the table given by the ref Man ( whatever the numbering from 0 to 7 or from 1 to 8 is).
For example on L476:
 the SPI1_TX  on Channel 3 (from 1 to 8) has request 1 (from 0 to 7)
 the SPI1_RX on Channel 2 (from 1 to 8) has request 1 (from 0 to 7) 
For example on F411:
 the SPI1_TX  on stream 5 (from 0 to 7) has request channel 3 (from 0 to 7)  
 the SPI1_RX on stream 2 (from 0 to 7) has request channel  3 (from 0 to 7)

File: include/dt-bindings/dma/stm32_dma.h (hunk 8 above)
Comment (FRASTM) [reply include/dt-bindings/dma/stm32_dma.h:28]: yes or No, I did not find where this is used. This field is valid for V1 only.
Ok, I let it unchanged

File: drivers/dma/dma_stm32_v2.c (hunk 10 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32_v2.c:120]: func_ll_clear_gi() function is used in the dma_stm32_irq_handler() of the dma_stm32.c

File: include/dt-bindings/dma/stm32_dma.h (hunk 8 above)
Comment (avisconti) [reply include/dt-bindings/dma/stm32_dma.h:28]: I prefer your new way! It seems cleaner.

File: drivers/dma/dma_stm32.c (hunk 11 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32.c:290]: We control here that the buffer for receiving or sending data is not 0.
 In case it is 0, it means the transfer will read (or write) mem address 0x0000 0000, which is not actually an error in terms of DMA memory transfer. 
Yes, this can be moved above.

File: drivers/dma/dma_stm32_v2.c (hunk 10 above)
Comment (avisconti) [reply drivers/dma/dma_stm32_v2.c:120]: OK, I missed that one! Sorry.

File: drivers/dma/dma_stm32.c (hunk 12 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32.c:338]: changed to` LOG_ERR("Channel direction error (%d)."`

File: include/dt-bindings/dma/stm32_dma.h (hunk 8 above)
Comment (avisconti) [reply include/dt-bindings/dma/stm32_dma.h:28]: So I agree with your changes.

File: drivers/dma/dma_stm32.h (hunk 9 above)
Comment (avisconti) [reply drivers/dma/dma_stm32.h:44]: YEs, very confusing.

File: drivers/dma/dma_stm32.c (hunk 11 above)
Comment (avisconti) [reply drivers/dma/dma_stm32.c:290]: Wait a moment.
I think that &config->head_block->source_address is always != from 0, isn't it?
THis is the address of a field in a structure. so != from 0.
I guess that check may be only:
if (config->head_block->source_address != 0)

What do you think?

File: drivers/dma/dma_stm32.c (hunk 12 above)
Comment (avisconti) [reply drivers/dma/dma_stm32.c:338]: ack

File: drivers/dma/dma_stm32.h (hunk 9 above)
Comment (avisconti) [reply drivers/dma/dma_stm32.h:44]: OK, I think it is correct in this way.

File: include/dt-bindings/dma/stm32_dma.h (hunk 8 above)
Comment (FRASTM) [reply include/dt-bindings/dma/stm32_dma.h:28]: do you mean 
`#define STM32_DMA_CONFIG_PERIPHERAL_INC_FIXED(config)	((config >> 15) & 0x1)` 

or 
```
#ifdef CONFIG_DMA_STM32_V1
/* priority increment offset defined on bit 15 */
#define STM32_DMA_CONFIG_PERIPH_INC_FIXED(config)	(config & 0x1 << 15)
#endif /* CONFIG_DMA_STM32_V1 */
```

File: drivers/dma/dma_stm32.c (hunk 11 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32.c:290]: That was the purpose of the testcase: `spi_null_tx_buf` in tests/drivers/spi/spi_loopback
to assign set 0 for the transmit (origin) mem address. and the structure tx_bufs is valid.
Here is what I want to check.

File: include/dt-bindings/dma/stm32_dma.h (hunk 8 above)
Comment (avisconti) [reply include/dt-bindings/dma/stm32_dma.h:28]: I mean the complete block of changes with the ((config >> 15) & 0x1) style.

File: drivers/dma/dma_stm32.c (hunk 11 above)
Comment (avisconti) [reply drivers/dma/dma_stm32.c:290]: OK, so do you agree that if (config->head_block->source_address != 0) should be enough?

File: include/dt-bindings/dma/stm32_dma.h (hunk 8 above)
Comment (FRASTM) [reply include/dt-bindings/dma/stm32_dma.h:28]: ok

File: drivers/dma/dma_stm32.c (hunk 11 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32.c:290]: yes, agreed

**Comment by erwango** [comment]
@avisconti, I've updated the PR with the 2 agreed changes

**Review by avisconti — APPROVED** [review]
LGTM

File: include/dt-bindings/dma/stm32_dma.h [hunk 13]
```diff
@@ -1,16 +1,38 @@
...
 #define STM32_DMA_CONFIG_PRIORITY(config)		((config >> 16) & 0x3)
 
 /* macros for features */
+#ifdef CONFIG_DMA_STM32_V1
```
Comment (galak) [inline include/dt-bindings/dma/stm32_dma.h:32]: We should avoid CONFIG defines in this header if at all possible.

File: dts/bindings/dma/st,stm32-dma.yaml [hunk 14]
```diff
@@ -23,15 +23,67 @@ properties:
...
+        -bit 15: Peripheral Increment Offset Size (not USED for DMA V2)
+       	       0x0: offset size is linked to the peripheral bus width
+       	       0x1: offset size is fixed to 4 (32-bit alignment)
+        -bit 16-17 : Priority level
+       	       0x0: low
+       	       0x1: medium
+       	       0x2: high
+       	       0x3: very high
+    4. features: A 32bit bitfield value specifying DMA features
+        -bit 0-1 or not used: DMA FIFO threshold selection
```
Comment (galak) [inline dts/bindings/dma/st,stm32-dma.yaml:76]: Not sure I follow here.  What does 'or not used' mean?

File: include/dt-bindings/dma/stm32_dma.h (hunk 13 above)
Comment (erwango) [reply include/dt-bindings/dma/stm32_dma.h:32]: ok, this can be added.

File: dts/bindings/dma/st,stm32-dma.yaml (hunk 14 above)
Comment (erwango) [reply dts/bindings/dma/st,stm32-dma.yaml:76]: Depending on the IP, it might not be used (this is the case for _v1)

**Review by galak — COMMENTED** [review]
a few things around Device tree changes.

File: dts/bindings/dma/st,stm32-dma.yaml (hunk 14 above)
Comment (erwango) [reply dts/bindings/dma/st,stm32-dma.yaml]: This was actually not used in V2 case. Added a clarification on that.

File: include/dt-bindings/dma/stm32_dma.h (hunk 13 above)
Comment (erwango) [reply include/dt-bindings/dma/stm32_dma.h]: Removed the conditional part.
Difference is already dealt with inside the driver.

**Comment by erwango** [comment]
@galak, this should be ok now
